azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,273 @@
+# coding=utf-8
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from multiprocessing.sharedctypes import Value
+from knack.log import get_logger
+from azext_iot.operations import hub
+from azext_iot.sdk.iothub import device
+from azure.cli.core.azclierror import (
+    CLIInternalError,
+    InvalidArgumentValueError,
+    RequiredArgumentMissingError,
+)
+from azext_iot.common.shared import SdkType, JobStatusType, JobType, JobVersionType
+from azext_iot.common.utility import handle_service_exception, process_json_arg
+from azext_iot.operations.generic import _execute_query, _process_top
+from azext_iot.iothub.providers.base import IoTHubProvider, CloudError, SerializationError
+from azext_iot.operations.hub import iot_device_list, iot_device_show, iot_device_create, iot_device_delete, \
+    iot_device_twin_show, iot_device_twin_replace, iot_hub_configuration_create, iot_hub_configuration_list, \
```

Comment:
The problem with many top level functions defined in operations.hub is they are optimized for one-off command process usage rather than general re-use, making them inefficient to use without a discovery.get_target passthrough (as this method will try and look up the target IoT Hub) via login connection string (limited to key based auth), or other mechanism. In the past, as a shortcut to efficiency we split up certain operations defined in the module in two parts ...one that includes IoT Hub discovery and one that just needs the returned target dictionary. Here is an example for that https://github.com/Azure/azure-iot-cli-extension/blob/dev/azext_iot/operations/hub.py#L108. This is not the most optimal solution, and we've iterated on vNext patterns a few times since then, but it shows the gist of a way to not look up the IoT Hub every time you call an IoT Hub service operation.